common/client-core/src/client/statistics_control.rs

```diff
@@ -0,0 +1,174 @@
+    pub(crate) fn new(
+        reporting_address: Recipient,
+        client_stats_id: String,
+        client_type: String,
+        report_tx: InputMessageSender,
+    ) -> (Self, ClientStatsSender) {
+        let (stats_tx, stats_rx) = tokio::sync::mpsc::unbounded_channel();
+
+        let stats = ClientStatsController::new(client_stats_id, client_type);
+
+        (
+            StatisticsControl {
+                stats,
+                stats_rx,
+                reporting_address,
+                report_tx,
+            },
+            ClientStatsSender::new(stats_tx),
+        )
+    }
+
+    async fn report_stats(&mut self) {
+        let stats_report = self.stats.build_report();
+
+        if let Ok(report_bytes) = stats_report.try_into() {
+            let report_message = InputMessage::new_regular(
+                self.reporting_address,
+                report_bytes,
+                TransmissionLane::General,
+                None,
+            );
+            if let Err(err) = self.report_tx.send(report_message).await {
+                log::error!("Failed to report client stats: {:?}", err);
+            } else {
+                self.stats.reset();
+            }
+        } else {
+            log::error!("Failed to serialize stats report. This should never happen");
+        }
+    }
+
+    pub(crate) async fn run_with_shutdown(&mut self, mut shutdown: nym_task::TaskClient) {
+        log::debug!("Started StatisticsControl with graceful shutdown support");
+
+        let report_interval = Duration::from_secs(STATS_REPORT_INTERVAL_SECS);
+        let mut report_interval = tokio::time::interval(report_interval);
+        let snapshot_interval = Duration::from_millis(SNAPSHOT_INTERVAL_MS);
+        let mut snapshot_interval = tokio::time::interval(snapshot_interval);
+
+        loop {
+            tokio::select! {
+                stats_event = self.stats_rx.recv() => match stats_event {
+                        Some(stats_event) => self.stats.handle_event(stats_event),
+                        None => {
+                            log::trace!("StatisticsControl: shutting down due to closed stats channel");
+                            break;
+                        }
+                },
+                _ = snapshot_interval.tick() => {
+                    self.stats.snapshot();
+                }
+                _ = report_interval.tick() => {
+                    self.report_stats().await;
+                }
+                _ = shutdown.recv_with_delay() => {
+                    log::trace!("StatisticsControl: Received shutdown");
+                    break;
+                },
+            }
+        }
+        log::debug!("StatisticsControl: Exiting");
+    }
+
+    pub(crate) fn start_with_shutdown(mut self, task_client: nym_task::TaskClient) {
+        spawn_future(async move {
+            self.run_with_shutdown(task_client).await;
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    // use std::sync::Arc;
```

Comment:
if those tests are not run, why not just remove the dead code?

Reply:
Pinging @jmwample for this one